### PR TITLE
[WIP] - Torch optimize

### PR DESCRIPTION
**Pull request by peasant98**
This PR gives users the option to optimize with Pytorch's automatic differentiation engine. It's still WIP, but raising it now to see the progress of this branch in relation to master.

### CODE REVIEW

File: robotic_skin/calibration/transformation_matrix.py [hunk 1]
```diff
@@ -231,4 +243,46 @@ def parameters(self):
...
+        """
+        self.is_tensor = True
+        self.matrix = self.transformation_matrix(*self.params)
+
+    def numpy_(self):
+        """
+        converts transformation matrix to a numpy array in place.
+        """
+        self.is_tensor = False
+        self.matrix = self.transformation_matrix(*self.params)
```
Comment (watakandai) [inline robotic_skin/calibration/transformation_matrix.py]: I'm afraid these things do not work for the TransformationMatrix which results from the dot product of two other TransformationMatrix `T = Ti * Tj`, since it does no longer hold `self.params`

**Review by watakandai — CHANGES_REQUESTED** [review]

File: robotic_skin/calibration/transformation_matrix.py (hunk 1 above)
Comment (peasant98) [reply robotic_skin/calibration/transformation_matrix.py]: yea

File: robotic_skin/calibration/transformation_matrix.py [hunk 2]
```diff
@@ -160,16 +167,18 @@ def __mul__(self, T):
             Resulting transformation matrix from dot products
             of two transformation matrices
         """
-        new_matrix = np.dot(self.matrix, T.matrix)
+        new_matrix = self.mm_fnc(self.matrix, T.matrix)
```
Comment (watakandai) [inline robotic_skin/calibration/transformation_matrix.py:202]: What if they had different class type?
What would be the resulting new_matrix class type?

And is this in a correct position? I believe this should be below `T.matrix = new_matrix`
```python
        if self.is_tensor:
            T.tensor_()
```

**Review by watakandai — CHANGES_REQUESTED** [review]

File: robotic_skin/calibration/transformation_matrix.py (hunk 2 above)
Comment (peasant98) [reply robotic_skin/calibration/transformation_matrix.py:202]: updated.

File: examples/calibration/calibrate_imu_poses.py [hunk 3]
```diff
@@ -0,0 +1,276 @@
+#!/usr/bin/env python
```
Comment (watakandai) [inline examples/calibration/calibrate_imu_poses.py]: I moved the script under the directory called `robotic_skin/robotic_skin/examples/calibration/`

File: robotic_skin/calibration/loss.py [hunk 4]
```diff
@@ -25,7 +27,7 @@ def __init__(self):
 
     def __call__(self, x_estimated, x_target, axis=0):
         x = x_estimated - x_target
-        return np.mean(np.linalg.norm(x, axis=axis, ord=1))
+        return np.mean(np.sum(x))
```
Comment (watakandai) [inline robotic_skin/calibration/loss.py:29]: you forgot `np.abs()`

File: robotic_skin/calibration/loss.py [hunk 5]
```diff
@@ -50,8 +55,50 @@ def __call__(self, x_estimated, x_target, axis=0):
...
 
 class MeanAbsoluteLoss(Loss):
+    """
+    mean absolute loss
+    """
     def __init__(self):
         super().__init__()
 
     def __call__(self, x_estimated, x_target, axis=0):
         return mean_absolute_error(x_target, x_estimated)
```
Comment (watakandai) [inline robotic_skin/calibration/loss.py:64]: These classes must be able to handle both a singular array and multi-dimension arrays.

File: robotic_skin/calibration/transformation_matrix.py [hunk 6]
```diff
@@ -22,13 +23,31 @@ def __init__(self, theta=None, d=None, a=None, alpha=None):
...
+        if type(theta) == torch.Tensor:
+            if theta is None:
+                theta = torch.tensor(0.).double().cuda()
+            if d is None:
+                d = torch.tensor(0.).double().cuda()
+            if a is None:
+                a = torch.tensor(0.).double().cuda()
+            if alpha is None:
+                alpha = torch.tensor(0.).double().cuda()
+            temp_params = np.array([theta.cpu(), d.cpu(), a.cpu(), alpha.cpu()], dtype=float)
```
Comment (watakandai) [inline robotic_skin/calibration/transformation_matrix.py]: why do you have 2 `temp_params`

File: robotic_skin/calibration/kinematic_chain.py [hunk 7]
```diff
@@ -1,7 +1,8 @@
 import copy
 import numpy as np
 from typing import List
-from .transformation_matrix import TransformationMatrix as TM
+from robotic_skin.calibration.transformation_matrix import TransformationMatrix as TM
+import torch
```
Comment (watakandai) [inline robotic_skin/calibration/kinematic_chain.py]: Don't forget to add the modifications I made in the last PR.

**Review by watakandai — CHANGES_REQUESTED** [review]

File: robotic_skin/calibration/kinematic_chain.py (hunk 7 above)
Comment (peasant98) [reply robotic_skin/calibration/kinematic_chain.py]: yea, many are coming

File: robotic_skin/calibration/loss.py (hunk 4 above)
Comment (peasant98) [reply robotic_skin/calibration/loss.py:29]: ah, thx

Comment (peasant98) [reply robotic_skin/calibration/loss.py:29]: fixed

File: examples/calibration/calibrate_imu_poses.py (hunk 3 above)
Comment (peasant98) [reply examples/calibration/calibrate_imu_poses.py]: yea, going to delete this.

Comment (peasant98) [reply examples/calibration/calibrate_imu_poses.py]: yea, thx

File: robotic_skin/calibration/kinematic_chain.py (hunk 7 above)
Comment (peasant98) [reply robotic_skin/calibration/kinematic_chain.py]: all is updated.

File: robotic_skin/calibration/loss.py (hunk 5 above)
Comment (peasant98) [reply robotic_skin/calibration/loss.py:64]: yep, they can

**Comment by peasant98** [comment]
finally, the green check mark is there.

File: robotic_skin/calibration/error_functions.py [hunk 8]
```diff
@@ -197,7 +197,7 @@ def __call__(self, kinematic_chain, i_su):
         ------------
         i_su: int
             i_su th sensor
-        kinemaic_chain:
```
Comment (watakandai) [inline robotic_skin/calibration/error_functions.py:200]: Thanks :+1:  LOL

Comment (peasant98) [reply robotic_skin/calibration/error_functions.py:200]: yep

File: robotic_skin/calibration/transformation_matrix.py [hunk 9]
```diff
@@ -22,13 +23,34 @@ def __init__(self, theta=None, d=None, a=None, alpha=None):
...
+            params = np.array([theta, d, a, alpha], dtype=float)
+            # Only select provided keys (which are not None)
+            self.params = np.nan_to_num(params)
+            temp_params = self.params
+
         self.matrix = self.transformation_matrix(*self.params)
-        qx = pyqt.Quaternion(axis=[1, 0, 0], angle=self.params[3])
-        qz = pyqt.Quaternion(axis=[0, 0, 1], angle=self.params[0])
+        qx = pyqt.Quaternion(axis=[1, 0, 0], angle=temp_params[3])
+        qz = pyqt.Quaternion(axis=[0, 0, 1], angle=temp_params[0])
```
Comment (watakandai) [inline robotic_skin/calibration/transformation_matrix.py]: Why don't you first run
```
params = np.array([theta, d, a, alpha], dtype=float)
self.params = np.nan_to_num(params)
```

and then 
1. set quaternions
2. convert it to tensor. 

Then you don't really need such a long if statement and `temp_params`

**Review by watakandai — CHANGES_REQUESTED** [review]

File: robotic_skin/calibration/optimizer.py [hunk 10]
```diff
@@ -147,15 +159,240 @@ def _optimize(self, i_su: int):
         raise NotImplementedError
 
 
-class NNOptimizerBase(OptimizerBase):
-    def __init__(self, kinematic_chain, evaluator, data_logger, optimize_all):
-        super().__init__(kinematic_chain, evaluator, data_logger)
+class TorchOptimizerBase(IncrementalOptimizerBase):
```
Comment (watakandai) [inline robotic_skin/calibration/optimizer.py:162]: Can you add comments on what each algorithm does?

Comment (peasant98) [reply robotic_skin/calibration/optimizer.py:162]: yes

File: robotic_skin/calibration/transformation_matrix.py (hunk 9 above)
Comment (peasant98) [reply robotic_skin/calibration/transformation_matrix.py]: yea, not sure what I was doing here.

Comment (peasant98) [reply robotic_skin/calibration/transformation_matrix.py]: fixed.

File: robotic_skin/calibration/transformation_matrix.py (hunk 6 above)
Comment (peasant98) [reply robotic_skin/calibration/transformation_matrix.py]: fixed

File: robotic_skin/calibration/optimizer.py (hunk 10 above)
Comment (peasant98) [reply robotic_skin/calibration/optimizer.py:162]: done.

**Comment by peasant98** [comment]
merging once build passes.

**Review by watakandai — APPROVED** [review]
